discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,266 @@
+	for {
+		select {
+		case ev := <-conn.eventChan:
+			_, _, _, err := conn.Client.GetW(ev.Path)
+			if err != nil {
+				log.Warning(err)
+				conn.Deregister(service)
+				return
+			}
+		case <-time.After(time.Duration(service.TTL) * time.Second):
+			conn.Deregister(service)
+			return
+		}
+	}
+}
+
+func (conn ZooKeeper) eventChanCallBack(ev zk.Event) {
+	if ev.Type == zk.EventNodeDataChanged {
+		go func() {
+			conn.eventChan <- ev
+		}()
+	}
+}
+
+// ConfigHook is the hook to register with the ZooKeeper backend
+func ConfigHook(raw interface{}) (discovery.ServiceBackend, error) {
+	return NewZooKeeperConfig(raw)
+}
+
+func connection(addresses []string, cb zk.EventCallback) (*zk.Conn, error) {
+	c, _, err := zk.Connect(addresses, time.Second, zk.WithEventCallback(cb))
+	return c, err
+}
+
+// NewZooKeeperConfig creates a new service discovery backend for zookeeper
+func NewZooKeeperConfig(raw interface{}) (*ZooKeeper, error) {
+	zookeeper := &ZooKeeper{
+		Prefix:    "/containerpilot",
+		eventChan: make(chan zk.Event),
+	}
+	config := &struct {
+		Address string `mapstructure:"address"`
+	}{}
+	if err := utils.DecodeRaw(raw, &config); err != nil {
+		return nil, err
+	}
+	conn, err := connection([]string{config.Address}, zookeeper.eventChanCallBack)
+	if err != nil {
+		return nil, err
+	}
+	zookeeper.Client = conn
+	return zookeeper, nil
+}
+
+// Deregister removes this instance from the registry
+func (conn *ZooKeeper) Deregister(service *discovery.ServiceDefinition) {
+	key := conn.getNodeKey(service)
+	exists, _, err := conn.Client.Exists(key)
+	if err != nil {
+		log.Warnf("Error on deregistering %s. Node should exists: %s", key, err)
+	}
+	if exists {
+		log.Warnf("Deregistering %s", key)
+		conn.Client.Delete(conn.getNodeKey(service), -1)
+	}
+}
+
+// MarkForMaintenance removes this instance from the registry
+func (conn *ZooKeeper) MarkForMaintenance(service *discovery.ServiceDefinition) {
+	conn.Deregister(service)
+}
+
+// SendHeartbeat refreshes the associated zookeeper node by
+// re-registering it.
+func (conn *ZooKeeper) SendHeartbeat(service *discovery.ServiceDefinition) {
+	if err := conn.registerService(service); err != nil {
+		log.Warnf("Error registering service %s: %s", service.Name, err)
+	}
+}
+
+func (conn *ZooKeeper) parentPath(service *discovery.ServiceDefinition) string {
+	return fmt.Sprintf("%s/%s", conn.Prefix, service.Name)
+}
+
+func (conn *ZooKeeper) getNodeKey(service *discovery.ServiceDefinition) string {
+	return fmt.Sprintf("%s/%s", conn.parentPath(service), service.ID)
+}
+
+func (conn *ZooKeeper) getAppKey(appName string) string {
+	return fmt.Sprintf("%s/%s", conn.Prefix, appName)
+}
+
+var zookeeperUpstreams = make(map[string][]ServiceNode)
+
+// CheckForUpstreamChanges checks another zookeeper node for changes
+func (conn *ZooKeeper) CheckForUpstreamChanges(backendName, backendTag string) bool {
+	// TODO: is there a way to filter by tag in zookeeper?
```

Comment:
It doesn't look like there's such a concept in ZK. We should remove this note and the `Tags` field of the results struct.